Module to generate Diffie-Hellman parameters

thomwiggers opened the pull request:
Fixes #32577

##### SUMMARY

Implements diffie-hellman parameter generation to complement the openssl crypto modules. Directly calls openssl due to pyopenssl not having a dhparam api.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
openssl_dhparam

##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
- hosts: localhost
  become: no
  tasks:
    - openssl_dhparam: size=1024 path=/tmp/dhparm
    - openssl_dhparam: path=/tmp/test state=absent
```

thomwiggers committed:
Module to generate Diffie-Hellman parameters

Implements #32577

thomwiggers commented:
I'm directly calling `openssl dhparam` as `pyOpenSSL` doesn't support DH parameter generation.

An alternative would be using the `cryptography` library but that would introduce yet another dependency.

MarkusTeufelberger commented:
Looks nice already! :+1: 

Please still use pyopenssl, if at all possible. You can call functions not offered by pyopenssl via the `OpenSSL` object like this: https://github.com/ansible/ansible/pull/30338/commits/793161b6b21120cc726147b6bac4c42adf4c61f1

In the documentation of return values you refer to private keys.

thomwiggers commented:
Using the pyopenssl-api appears to be quite complicated. I'm trying to follow https://wiki.openssl.org/index.php/Diffie_Hellman#Using_the_Low_Level_APIs but having trouble caling `DH_generate_parameters_ex` as I can't find `DH_generator_2` or  how to create the (null) `BN_GENCB *` pointer.

Using the pyca/cryptography module would be much more convenient (possibly also for the other modules)... but brings in new dependencies.


I've fixed the mentions of private keys.

thomwiggers commented:
It may actually be desirable to refactor all modules to use the cryptography library; it'd make implementing #32626 a lot easier as well.

MarkusTeufelberger commented:
I fully agree, however it was a hard requirement for my colleague and me when writing openssl_certificate and refatoring openssl_csr that it has to run on Debian Jessie. We absolutely need openssl_privatekey, openssl_csr and openssl_certificate to be fully functional on Jessie hosts. This realistically only leaves a relatively old version of pyopenssl.

Once Jessie is out of support (April 2020) I'm planning to take another look at cryptography.

thomwiggers commented:
Won't this just work: https://packages.debian.org/jessie/python-cryptography ? I haven't tried it, but it seems (some) version of cryptography is available.

MarkusTeufelberger commented:
Not surprisingly, since cryptography is a dependency of pyopenssl. :-)

Unfortunately, no, this won't just work. Certificate and CSR handling were implemented after 0.6 and reworked a few times later.

MarkusTeufelberger commented:
Well, since pyopenssl does not seem like an option for now and I have no need for that module anyways, I'd just like you to use `extends_documentation_fragment: files` as in #32664 (and make sure that at least the mode of the DH params file is set to something reasonable right from the start), then you can have my ok.

Essentially just calling out to `openssl` is definitely not a nice solution, but if it works for you, who am I for keeping you from submitting something that you need and that works for you upstream.

Lastly it would be really great if you also added at least some small integration tests (take a look at the existing ones) in a future PR, this stuff has a tendency of being a bit difficult on various platforms and having at least some basic assertion that something works as intended somewhere is always helpful.

MarkusTeufelberger commented:
shipit

thomwiggers commented:
I'm still adding some integration tests. They're currently running on my machine, unfortunately generating 2048 bit keys takes a while...

MarkusTeufelberger commented:
Great! :)

Maybe just generate 1024 bit keys then, it's anyways more about making sure that limits are respected than to create something that would be used 1:1 in production.

thomwiggers committed:
Add integration tests for openssl_dhparam

mattclay commented:
CI failure in integration tests:

- osx: https://app.shippable.com/github/ansible/ansible/runs/44867/11/tests (freebsd 10.3 and 11.0 fail the same way)
- centos6: https://app.shippable.com/github/ansible/ansible/runs/44867/14/tests (fails with a traceback)

thomwiggers commented:
The FreeBSD errors are due to libressl, I suspect, and the Centos issue is due to python 2.6 not having check_output. I'm adding a backported check_output and currently am debugging the output differences.

MarkusTeufelberger commented:
shipit

MarkusTeufelberger commented:
shipit

thomwiggers commented:
bot_status

thomwiggers commented:
So I guess that means I need a `shipit` from @Spredzy ?

MarkusTeufelberger commented:
I guess so, I've usually had luck asking on IRC for someone to take a look.

mscherer commented on lib/ansible/modules/crypto/openssl_dhparam.py:
```diff
@@ -0,0 +1,252 @@
+        """Generate a keypair."""
+        # openssl dhparam -out <path> <bits>
+        if not self.force and self.check(module):
+            return
```
I suspect there is a subtle bug lurking here. If I generate a dhparam file, and then later changes permissions, the return will prevent the permission to be applied, since that's done at the end of the function and only if we did generate a param.

thomwiggers replied:
`check()` will return `False` is the attributes (permissions) do not match; this will make it pass this check and then change the permissions (and re-generate parameters in this case – perhaps I should refactor it to only do that if the bits don't match).

mscherer replied:
ok, as pointed on irc, that's verified by check, so that's ok.

mscherer commented:
shipit

thomwiggers committed:
Slightly refactor check to prevent unnecessary regeneration

thomwiggers commented:
I've refactored `check()` and `generate()` to not regenerate the full dh parameters when only changing permissions.

thomwiggers commented:
Sorry for that noise. It would be appreciated if the PR could be reviewed again. I think the last commit should make what's going on a lot more clear.

MarkusTeufelberger commented:
shipit

Spredzy commented:
Hey @thomwiggers thanks for submitting this PR.

As a personal feeling I'd rather like we avoid using `openssl` directly in the modules.

`pyOpenSSL` was used for `openssl_privatekey`, `openssl_publickey`, `openssl_csr` and `openssl_certificate` because if offered an easy interface to interact with. Nothing actually prevent from using `cryptography` directly if it is the most obvious way for what the module is trying to achieve.

Hence, and sorry for the delayed answer, if @MarkusTeufelberger has nothing against it I'd rather like we go forward with `cryptography` to generate and validate `openssl_dhparam` module has it will make the code way cleaner.

Generation looks simple (example):

```
parameters = dh.generate_parameters(generator=GENERATOR, key_size=SIZE, backend=default_backend())

parameters_serialized = parameters.parameter_bytes(Encoding.PEM, ParameterFormat.PKCS3)

with open(PATH, 'w') as f:
     f.write(parameters_serialized)
```

So does verification (example):

```
key = load_pem_parameters(open(PATH).read(), default_backend())
return key.parameter_numbers().p.bit_length() == SIZE and key.parameter_numbers().g == GENERATOR
```

The only caveat is that `ParameterFormat.PKCS3` was introduced in `cryptography==2.0`. This version is shipped in latest Debian, Fedora, FreeBSD (those are the one I checked). Per-se it is not an issue and this shouldn't prevent us from using it. Like for other modules in this name space you should put a try/catch around the `parameter_bytes()` call stating `cryptography 2.0 is required`. (Same thing is done for pyOpenSSL>1.5 in `openssl_publickey`, `openssl_certificate` and `openssl_csr`)

Spredzy reviewed (changes requested):
See above comment

thomwiggers commented:
In Ubuntu 16.04, the package `python-cryptography` is still stuck at version 1.2.3, unfortunately. 

What's also weird, even though the docs claim support for `dh.generate_parameters` starting from 0.9, this happens:

```python
>>> cryptography.__version__
'1.2.3'
>>> from cryptography.hazmat.primitives.asymmetric import dh
>>> dh.generate_parameters(generator=2, key_size=512, backend=default_backend())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'generate_parameters'
```

I think this was really only supported starting from 1.7: https://github.com/pyca/cryptography/issues/4067

MarkusTeufelberger commented:
I would be all for a rewrite using latest py-cryptography of all crypto modules... than again, for csr, cert and privkey I need support on Debian Jessie, if necessary with backports.

Anyways, I don't need this module, so I don't have anything against writing it for py-cryptography instead of calling out to the openssl binary on the system. Maybe integration tests should more clearly spell out if a module is even supported on that platform though, so it would be easier to find out who is affected.

Spredzy commented:
@thomwiggers would that prevent you from doing something you need that this module wouldn't work on Ubuntu 16.04 out-of-the-box ?

thomwiggers commented:
Hm, it would be quite annoying to have to install python-cryptography from pip for my current deployments. The package cryptography ≥ 2.0 isn't in any Ubuntu before 18.04 either, as of right now – but on the other hand I suppose 18.04 will be available and stable soonish – maybe not that soon after or before Ansible 2.5 is released.

If only ansible could set up a virtualenv to work in on the remote host :stuck_out_tongue:.

I think using `cryptography` would probably work out for my use cases as those should mostly be on Ubuntu 18.04 soon. The current `dhparam` could probably be moved to Ansible Galaxy as for 'legacy' systems as well. Maybe moving "old" stuff into Galaxy is also a solution for cleaning up the other `openssl_` modules?

As a more general note, it's slightly confusing that the Ansible modules don't come with clear targeting/guidelines for minimal support: it would clear up a lot of headaches like this.

MarkusTeufelberger commented:
> As a more general note, it's slightly confusing that the Ansible modules don't come with clear targeting/guidelines for minimal support: it would clear up a lot of headaches like this.

I agree, but that's probably better discussed on a mailing list. :-)

thomwiggers commented:
As an aside, once you release something requiring at least (e.g.) Ubuntu 18.04, it's a matter of time before issues like https://github.com/ansible/ansible/issues/34054 pop up – at least in the current state of this module it will practically run anywhere (I'm not sure on the state of cryptography on OS X btw).

Spredzy commented:
Overall agree with what has been said here. The issue here is that we deal with distribution and have no way to control which version of the package they will shipped with. So yes #34054 is inevitable. This is something that has been 'accepted' the day `pyOpenSSL>=15.0` has been made a hard requirement.

On the other hand using the openssl command directly doesn't require us really to create a module for that. User can put a properly crafted role/playbook that will make generation idempotent and move along.

So I am a bit torn here. Yes supporting as many OSes as possible out of the box would be ideal. But at which maintenance cost? As the modules are written today, we 'trust' `cryptography` and/or `pyOpenSSL` module version.

MarkusTeufelberger commented:
Maybe we should do 2 parallel implementations of all the crypto/ modules? The openssl_ones and then some with the same interface that require a very recent version of py-cryptography? Sounds like yet another maintenance nightmare though, but eventually these modules will need to be rewritten anyways as pyopenssl is being deprecated upstream. Might be good to wait for some guidance and tooling from the Ansible project though (e.g. integration tests that check for installed version and report compatibility).

One of the main issues is anyways that "agentless" is a lie for some/many modules.

thomwiggers commented:
> On the other hand using the openssl command directly doesn't require us really to create a module for that. User can put a properly crafted role/playbook that will make generation idempotent and move along.

One important thing this module has over just using `command` with `creates` is that it will validate validity and size of generated dh parameters: this is annoying to do by hand (see also the integration tests I've committed).

Spredzy commented:
I don't want to block it further, as it works in its current state.

I'll ask in a public core meeting what we should target (lib vs. os) and get back to you with the answer.

Anyway thanks @thomwiggers for this contribution and the exchanges, much apprecatiated.

shipit

thomwiggers commented:
If the eventual concensus is to rewrite to target a sufficiently capable version of python-cryptography, let me know, I'd love to help.

thomwiggers committed:
Fix code smell in tests

Highly annoying to have to do this again and again and again as the rules change during the game

thomwiggers commented:
┻━┻ ︵ヽ(`Д´)ﾉ︵﻿ ┻━┻

I've fixed this new 'code smell'. I've also taken the opportunity to rebase on devel – but I guess this means we need new shipits :(

MarkusTeufelberger commented:
shipit

Spredzy commented:
shipit

felixfontein commented on lib/ansible/modules/crypto/openssl_dhparam.py:
```diff
@@ -0,0 +1,269 @@
+            try:
+                # openssl dhparam -out <path> <bits>
+                subprocess.check_call(
+                    ["openssl", "dhparam", "-out", self.path, str(self.size)])
```
Out of curiosity: is there a reason why you aren't using `module.run_command()`, and `module.get_bin_path('openssl', True)` to get the path to the `openssl` binary?

thomwiggers replied:
Mainly because I had no idea that even existed.

felixfontein commented on lib/ansible/modules/crypto/openssl_dhparam.py:
```diff
@@ -0,0 +1,269 @@
+                subprocess.check_call(
+                    ["openssl", "dhparam", "-out", self.path, str(self.size)])
+            except CalledProcessError as exc:
+                os.remove(self.path)
```
Wouldn't it be better to let `openssl dhparam` write into a temp file, and on success move the temp file to the real file (with `module.atomic_move()`)? Then in case of interruptions or errors, existing destinations wouldn't be trashed (except of course if `atomic_move` itself goes terribly wrong).

felixfontein commented:
I've tried adding `module.run_command()` and writing into a tempfile; you can see the result here: https://gist.github.com/felixfontein/af3944b40228f8333a93ad604a5fe878

thomwiggers commented:
It's a nice solution, aside from some minor issues (you need to close the file handle returned by `tempfile.mkstemp`). However, it's quite hard for me to merge it as it's only a diff and not a valid patch file. Could you submit a patch file or create a pull request on my fork to this branch? That'll allow me to review it better.

felixfontein commented:
That's why I included the whole file in the gist as well, so you can copy it over and do `git diff` yourself :)
I'll try to create a PR on your branch.

felixfontein committed:
Using module.run_command() and module.atomic_move() from a tempfile.

felixfontein commented:
I haven't managed to create a PR in a sane way. You can try it yourself from https://github.com/felixfontein/ansible/tree/thomwiggers_openssl_dhparam.

Or try this:
``` .patch
diff --git a/lib/ansible/modules/crypto/openssl_dhparam.py b/lib/ansible/modules/crypto/openssl_dhparam.py
index 58d3109568..4f0e5045c4 100644
--- a/lib/ansible/modules/crypto/openssl_dhparam.py
+++ b/lib/ansible/modules/crypto/openssl_dhparam.py
@@ -82,48 +82,7 @@ filename:
 
 import os
 import re
-
-"""Note: We also monkey-patch subprocess for python 2.6 to
-give feature parity with later versions.
-Monkey-patch from http://pydoc.net/pep8radius/0.9.0/ with slight modifications
-MIT Licence
-
-Remove patch when Python 2.6 is no longer supported
-"""
-import subprocess
-try:
-    from subprocess import check_output, CalledProcessError
-except ImportError:  # pragma: no cover
-    # python 2.6 doesn't include check_output
-    # monkey patch it in!
-    def check_output(*popenargs, **kwargs):
-        if 'stdout' in kwargs:  # pragma: no cover
-            raise ValueError('stdout argument not allowed, '
-                             'it will be overridden.')
-        process = subprocess.Popen(stdout=subprocess.PIPE,
-                                   *popenargs, **kwargs)
-        output = process.communicate()[0]
-        retcode = process.poll()
-        if retcode:
-            cmd = kwargs.get("args")
-            if cmd is None:
-                cmd = popenargs[0]
-            raise CalledProcessError(retcode, cmd, output=output)
-        return output
-
-    # overwrite CalledProcessError due to `output`
-    # keyword not being available (in 2.6)
-    class CalledProcessError(Exception):
-        """Exception raised for call errors"""
-
-        def __init__(self, returncode, cmd, output=None):
-            self.returncode = returncode
-            self.cmd = cmd
-            self.output = output
-
-        def __str__(self):
-            return "Command '%s' returned non-zero exit status %d" % (
-                self.cmd, self.returncode)
+import tempfile
 
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils._text import to_native
@@ -141,6 +100,7 @@ class DHParameter(object):
         self.size = int(module.params['size'])
         self.force = module.params['force']
         self.changed = False
+        self.openssl_bin = module.get_bin_path('openssl', True)
 
     def generate(self, module):
         """Generate a keypair."""
@@ -148,15 +108,20 @@ class DHParameter(object):
 
         # ony generate when necessary
         if self.force or not self._check_params_valid(module):
+            # create a tempfile
+            f, tmpsrc = tempfile.mkstemp()
+            f.close()
+            module.add_cleanup_file(tmpsrc)  # Ansible will delete the file on exit
+            # openssl dhparam -out <path> <bits>
+            command = [self.openssl_bin, 'dhparam', '-out', tmpsrc, str(self.size)]
+            rc, _, err = module.run_command(command, check_rc=False)
+            if rc != 0:
+                raise DHParameterError(to_native(err))
             try:
-                # openssl dhparam -out <path> <bits>
-                subprocess.check_call(
-                    ["openssl", "dhparam", "-out", self.path, str(self.size)])
-            except CalledProcessError as exc:
-                os.remove(self.path)
-                raise DHParameterError(str(exc))
-            finally:
-                changed = True
+                module.atomic_move(tmpsrc, self.path)
+            except Exception as e:
+                module.fail_json(msg="Failed to write to file %s: %s" % (self.path, str(e)))
+            changed = True
 
         # fix permissions (checking force not necessary as done above)
         if not self._check_fs_attributes(module):
@@ -174,12 +139,10 @@ class DHParameter(object):
 
     def _check_params_valid(self, module):
         """Check if the params are in the correct state"""
-        try:
-            result = to_native(check_output(
-                ["openssl", "dhparam", "-check", "-text", "-noout", "-in", self.path],
-                stderr=subprocess.STDOUT,
-            )).strip()
-        except CalledProcessError as e:
+        command = [self.openssl_bin, 'dhparam', '-check', '-text', '-noout', '-in', self.path]
+        rc, out, err = module.run_command(command, check_rc=False)
+        result = to_native(out)
+        if rc != 0:
             # If the call failed the file probably doesn't exist or is
             # unreadable
             return False
@@ -191,7 +154,7 @@ class DHParameter(object):
             bits = int(match.group(1))
 
         # if output contains "WARNING" we've got a problem
-        if "WARNING" in result:
+        if "WARNING" in result or "WARNING" in to_native(err):
             return False
 
         return bits == self.size

```

felixfontein commented:
(And not closing the handle returned by `tempfile.mkstemp` is then a bug in the `letsencrypt` module as well -- that's where I copied the tempfile creation from.)

felixfontein commented:
(The `f.close()` should be `os.close(f)`...)

thomwiggers committed:
Remove underscore variable

Ansible prefers dummy

thomwiggers commented:
Ok, that does mean we need to enter the review process all over again :crying_cat_face:

felixfontein commented on lib/ansible/modules/crypto/openssl_dhparam.py:
```diff
@@ -114,7 +114,7 @@ def generate(self, module):
             # openssl dhparam -out <path> <bits>
             command = [self.openssl_bin, 'dhparam', '-out', tmpsrc, str(self.size)]
-            rc, _, err = module.run_command(command, check_rc=False)
+            rc, _out, err = module.run_command(command, check_rc=False)
```
Do you plan to use `_out`?

thomwiggers replied:
No, ansible doesn't allow `_` as a variable name for reasons unknown.

felixfontein replied:
Are you sure? `_` has been in use in the letsencrypt module (see https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/web_infrastructure/letsencrypt.py) for a long time, and so far nobody complained. (Maybe it's not alloewd in `jinja2` templates?)

thomwiggers replied:
I think it's a new thing in the CI checks. It's the reason https://app.shippable.com/github/ansible/ansible/runs/51995/summary failed though.

thomwiggers replied:
See also https://github.com/ansible/ansible/pull/32620#issuecomment-346000767

felixfontein replied:
Very interesting. I haven't seen this happening for all my `letsencrypt` PRs over the last few weeks. I'll investigate a bit, this made me very curious :)

felixfontein replied:
I found something: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/testing/sanity/no-underscore-variable.rst
Interesting... So the recommended name is `dummy`. I'll create another PR for `letsencrypt`...

felixfontein replied:
Would you mind renaming `_out` to `dummy`?

thomwiggers replied:
I like `_out` slightly better, it at least still documents that it's the stdout of the command while the `_`  indicates it's not used. Also it would again reset all the shipits to 0 :disappointed:

MarkusTeufelberger replied:
I regularly check it anyways and you need one from the core team, you can ping people on IRC in the #ansible-devel channel.

I like neither _out nor dummy, but `dummy` seems to be the one that the ansible project prefers...

thomwiggers replied:
Ok, ok, done it. I think core team will only review it after it's marked `shipit`

felixfontein commented:
Well, the code got shorter now ;-) Also, the change isn't that complicated, so it should go faster. And it's still two weeks until community module freeze for Ansible 2.5, that should be more than enough.

thomwiggers commented:
bot_status

MarkusTeufelberger commented:
shipit

MarkusTeufelberger commented:
Though you need one from @Spredzy too I think. ;-)

Also it probably needs to be manually merged since it contains changes to several files in different folders and ansibot doesn't merge that automatically afair.

thomwiggers commented:
According to the workflow documentation new modules always need a core team member to review them.

felixfontein reviewed (approved):
LGTM shipit

MarkusTeufelberger commented:
shipit

felixfontein commented:
bot_status

gundalow commented:
rebuild_merge

mscherer commented:
shipit

felixfontein commented:
@gundalow: it looks like one of the CI processes is hanging (https://app.shippable.com/github/ansible/ansible/runs/52245/45/console)

thomwiggers commented:
bot_status
